Core/Src/main.c

```diff
@@ -87,17 +91,37 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
   /* USER CODE BEGIN 2 */
-
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);  // Start PWM on channel 1
+  uint8_t rxData[3];
+  uint8_t txData[3];
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   while (1)
   {
+	// Set the chip select low to start communication
+	HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_RESET);
+
+	txData[0] = 0x1;  // 00000001 for start bit
+	txData[1] = 0x80; // 10000000 for single-ended on CH1
+
+	HAL_SPI_TransmitReceive(&hspi1, txData, rxData, 3, HAL_MAX_DELAY);
+	uint16_t adc_value = ((rxData[1] << 8) & (0x3 << 8)) | rxData[2];
+
+	// Set the chip select high to end communication
+	HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_SET);
+
+	uint16_t pwm_value = (adc_value / ADC_MAX_VALUE)/DUTY_CYCLE_RANGE + DUTY_CYCLE_RANGE;
```

Comment:
You are almost there- You could say Duty cycle range + (Duty cycle range * (adc_value/ADC MAX VALUE))

Reply:
yeah idk why i divided by duty cycle range i meant to multiply by it, i think i was just low on sleep lol